### PR TITLE
Fix schema/type inference issue #216

### PR DESCRIPTION
**Pull request by tanwanirahul**
The issue with current implementation is that, it bumps up the NullType to StringType in reduce phase of each partition. When the final reduce across partitions takes place the column type is inferred as StringType even though the column has only one integer value because of incorrect bumping up of NullTypes.

Here, in this pull, we don't bump up the Null to String as a part of reduce operation (within and across partitions). After the reduction, when we create Fields of StructType, we convert NullTypes to StringTypes.

### CODE REVIEW

**Comment by falaki** [comment]
Thanks a lot for submitting this. The code looks good. Would you add another unit test that verify it works end-to-end. Maybe use the input you provided in the issue:
```
A,B,C,D
1,,,
,1,,
,,1,
,,,1
```

**Comment by tanwanirahul** [comment]
Added an end to end test case to validate the proper schema/type inference. @falaki Take a look and merge if everything is fine.

File: src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala [hunk 1]
```diff
@@ -52,4 +83,14 @@ class InferSchemaSuite extends FunSuite {
       Array(LongType)).deep == Array(DoubleType).deep)
   }
 
+  test("Type/Schema inference works as expected for the simple parse dataset.")
```
Comment (HyukjinKwon) [inline src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala]: Hm.. isn't this a rather end-to-end test?

File: src/main/scala/com/databricks/spark/csv/CsvParser.scala [hunk 2]
```diff
@@ -117,12 +116,12 @@ class CsvParser extends Serializable {
     this
   }
 
-  /** Returns a Schema RDD for the given CSV path. */
```
Comment (HyukjinKwon) [inline src/main/scala/com/databricks/spark/csv/CsvParser.scala]: This is just my personal thought. This refactoring looks pretty reasonable in a way though.. Could we maybe do this in a separate PR?

Comment (tanwanirahul) [inline src/main/scala/com/databricks/spark/csv/CsvParser.scala]: Okay. Would make more sense.

File: src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala (hunk 1 above)
Comment (tanwanirahul) [inline src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala]: I agree that CsvSuite performs all end to end tests but since there was a dedicated suite for SchemaInference, I did prefer to put the schema tests in there. Do you see any issues in that?

Comment (HyukjinKwon) [inline src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala]: No, I think it is okay though. I just said this because I see it has been added in this way.

For example, https://github.com/databricks/spark-csv/pull/235 and https://github.com/databricks/spark-csv/pull/224

Comment (tanwanirahul) [inline src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala]: Reverted the refactoring part. Tests I kept as is. If more people feel CsvSuite would rather be the right place to put these tests in, I will make that change.

**Comment by tanwanirahul** [comment]
@HyukjinKwon Created https://github.com/databricks/spark-csv/pull/248 for the refactoring part.

File: src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala [hunk 3]
```diff
@@ -40,6 +63,14 @@ class InferSchemaSuite extends FunSuite {
     assert(InferSchema.inferField(LongType, "2015-08 14:49:00") == StringType)
   }
 
+  test("Merging Nulltypes should yeild Nulltype.")
+  {
+      assert(
```
Comment (falaki) [inline src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala]: Nit: the indent is off:
```scala
assert(
  InferSchema.mergeRowTypes(Array(NullType),
  Array(NullType)).deep == Array(NullType).deep)
```

File: src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala [hunk 4]
```diff
@@ -52,4 +83,14 @@ class InferSchemaSuite extends FunSuite {
       Array(LongType)).deep == Array(DoubleType).deep)
   }
 
+  test("Type/Schema inference works as expected for the simple sparse dataset.")
+  {
+    val df = new CsvParser().withUseHeader(true).withInferSchema(true)
+            .csvFile(sqlContext, simpleDatasetFile)
+    assert(
+        df.schema.fields.map{field => field.dataType}.deep ==
```
Comment (falaki) [inline src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala]: Nit: use two spaces for indent. of this line. Next line's indent is fine.

File: src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala [hunk 5]
```diff
@@ -52,4 +83,14 @@ class InferSchemaSuite extends FunSuite {
       Array(LongType)).deep == Array(DoubleType).deep)
   }
 
+  test("Type/Schema inference works as expected for the simple sparse dataset.")
```
Comment (falaki) [inline src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala]: Please move this test to `CsvSuite.scala`. In that suite, there is already a test SQL context, and you don't need to construct SQLContext in this file.

File: src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala [hunk 6]
```diff
@@ -2,8 +2,31 @@ package com.databricks.spark.csv.util
...
+import com.databricks.spark.csv.CsvRelation
 
-class InferSchemaSuite extends FunSuite {
+class InferSchemaSuite extends FunSuite with BeforeAndAfterAll {
+
+  private val simpleDatasetFile = "src/test/resources/simple.csv"
+  private val utf8Charset = "utf-8"
+  private var sqlContext: SQLContext = _
+
+  override def beforeAll(): Unit =
```
Comment (falaki) [inline src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala]: These can be removed once you move your end-to-end test to `CsvSuite.scala`. This suite a unit test and does not need to create a SQLContext.

**Comment by falaki** [comment]
@tanwanirahul left some more comments. Once  you address I am going to merge this. After that would you mind submitting a patch to spark. Spark-csv is now inlined in Spark 2.0.

**Comment by tanwanirahul** [comment]
@HyukjinKwon Moved end to end tests to CSVSuite, @falaki Also recommended the same.

@falaki Take a look at the CSVSuite for test case and InferSchemaSuite for Indent comments you left. Also let me know your thoughts on https://github.com/databricks/spark-csv/pull/248 

I will push the patch to spark repo soon.

File: src/test/scala/com/databricks/spark/csv/CsvSuite.scala [hunk 7]
```diff
@@ -717,6 +717,16 @@ abstract class AbstractCsvSuite extends FunSuite with BeforeAndAfterAll {
 
     assert(results.size === numCars)
   }
+
+  test("Type/Schema inference works as expected for the simple sparse dataset.")
+  {
```
Comment (HyukjinKwon) [inline src/test/scala/com/databricks/spark/csv/CsvSuite.scala]: I believe this `{` should be move up to the previous line.

File: src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala [hunk 8]
```diff
@@ -55,6 +55,13 @@ class InferSchemaSuite extends FunSuite {
     assert(InferSchema.inferField(LongType, "2015-08 14:49:00") == StringType)
   }
 
+  test("Merging Nulltypes should yeild Nulltype.")
+  {
```
Comment (HyukjinKwon) [inline src/test/scala/com/databricks/spark/csv/util/InferSchemaSuite.scala]: Ditto.

**Comment by tanwanirahul** [comment]
@HyukjinKwon @falaki Is there any eclipse formatting profile that I could load and apply it for the code I am writing?

**Comment by tanwanirahul** [comment]
@falaki Could we re-build the commit? I don't think build failed because of the code issue.

**Comment by tanwanirahul** [comment]
@falaki @HyukjinKwon Could you please look at what is causing the build to fail? Looking at the build error, I don't think its code issue.

**Comment by tanwanirahul** [comment]
@falaki Could you please suggest next steps for this? Also, is there any date planned for the next release? We are currently using this change by adding an unmanaged dependency.

**Comment by falaki** [comment]
Would you please try that combination of spark and openjdk version to investigate what goes wrong?

**Comment by HyukjinKwon** [comment]
Let me take a look when I have some time.

**Comment by HyukjinKwon** [comment]
Hm.. It looks like this was failed due to the cached jars in travis.ci. This works fine both on my local and travis with a new build [here](https://travis-ci.org/HyukjinKwon/spark-csv/builds/108464686).

I will create a new PR for this. Please make the author @tanwanirahul  when you merge that PR by `merge_pr.py`.

**Comment by HyukjinKwon** [comment]
Can we maybe close this?
